pgjdbc/src/main/java/org/postgresql/hostchooser/HostChooserFactory.java

```diff
@@ -15,10 +20,30 @@
 public class HostChooserFactory {
 
   public static HostChooser createHostChooser(HostSpec[] hostSpecs,
-      HostRequirement targetServerType, Properties info) {
-    if (hostSpecs.length == 1) {
-      return new SingleHostChooser(hostSpecs[0], targetServerType);
+      HostRequirement targetServerType, Properties info) throws PSQLException {
+    if (hostSpecs.length != 1) {
+      return new MultiHostChooser(hostSpecs, targetServerType, info);
     }
-    return new MultiHostChooser(hostSpecs, targetServerType, info);
+    HostSpec host = hostSpecs[0];
+    if (!host.shouldResolve()) {
+      return new SingleHostChooser(host, targetServerType);
+    }
+    final InetAddress[] all;
+    try {
+      all = InetAddress.getAllByName(host.getHost());
+    } catch (UnknownHostException e) {
+      throw new PSQLException(GT.tr("The connection attempt failed."),
+          PSQLState.CONNECTION_UNABLE_TO_CONNECT, e);
+    }
+    if (all.length == 1) {
+      return new SingleHostChooser(host, targetServerType);
+    }
+    HostSpec[] resultHosts = new HostSpec[all.length];
+    for (int i = 0; i < all.length; i++) {
+      InetAddress inetAddress = all[i];
+      resultHosts[i] = new HostSpec(inetAddress.getHostAddress(), host.getPort(),
+          host.getLocalSocketAddress());
```

Comment:
I am afraid it would break TLS hostname verification.

We already support multi-host. What we miss is IP-based balancing.

Reply:
How does it work for libpq then ?

Reply:
I can put `InetAddress` in `HostSpec` and use it in `PGStream` (I believe this is the only place where actual connection is created), will it be ok?

Reply:
Sure we need to remember both "hostname" and "the resolved IP address".

Suppose there's "db1" which resolves to 3 IP addresses and "db2" which resolves to 15 IP addresses.
Does it mean we distribute the load as 3:15? Should we distribute the load as 1:1 and then make even distribution among the IPs behind each hostname?

I suggest we settle on that before adjusting the implementation.

Reply:
Tbh I don't think that this would be the case, more like "db1" that resolves to 18 ip addresses and that's it.
Right now implementation doesn't even bother to detect ip addresses if multiple hosts was defined.

Reply:
So, [libpq](https://github.com/postgres/postgres/blob/master/src/interfaces/libpq/fe-connect.c#L3967) looks at provided hosts as a single backend per host: if it connects to any address of provided host and finds out it is not in a desired state it rejects the host and don't even bother try next addresses.

Golangs [pgx](https://github.com/jackc/pgx/blob/master/pgconn/pgconn.go#L147) treats every address as a separate instance and tries them all, only stops on auth errors. That behavior would be more preferred for me as I have case for single hostname backed with addresses for every db instance.

Error propagations becomes more interesting anyway, because connecting to something like localhost will throw some interesting exceptions like `Could not find a server with specified targetServerType: any` instead of auth errors, but it can be fixed.

Reply:
@vlsi @davecramer so which implementation should I pick as reference.

Reply:
actually, cloud providers do this quite a bit. Where there is 1 writer and N readers. Active-Active is gaining traction in PostgreSQL so having N writers and M readers will be coming. If I had to choose between the two implementations it would probably be libpq only because more interfaces use it.